### ISSUE TITLE
lxc exec: unexpected stderr redirect behaviour in the interactive mode

### ISSUE THREAD

**Issue by dmitry-lyfar** [canonical/lxd#12326]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 22.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.19.0-50-generic
   * LXC version: 5.0.0~git2209-g5a7b9ce67
   * LXD version: lxd   git-fd28846  25670  latest/edge
   * Storage backend in use: dir

# Issue description

If LXD Exec runs a command in the interactive mode, the inability to separate stderr and stdout in the API leads to quirky issues with the command output.

# Steps to reproduce

 1. `lxc exec u1 --user 1000 --group 1000 -- ls -R / 2>/dev/null`
 2. One can still see something like that in the output `ls: cannot open directory '/var/spool/cron/atjobs': Permission denied`
 3. It's not possible to filter out stderr here unless the exec is enforced with -T. The reason is that LXD client sees that both stdin and stdout are connected to the terminal, starts the command in the interactive mode and combines stdin and stderr on the server as a result. One has to know about this LXD API Exec intricacy; otherwise they would start questioning their sanity.

**Comment by tomponline** [comment]
I'm not sure we can fix this. But the redirection you've provided in 1. is actually being applied on the client side not on the server side. As you say correctly there is only one output stream with interactive mode, so you would need to filter on the server side using an invocation like this:

```
lxc exec c1 -t -- bash -c "hostname > ~/foo.txt"
```

**Comment by tomponline** [comment]
@dmitry-lyfar Please let me know if this answers your question and we can close this. Thanks

**Comment by dmitry-lyfar** [comment]
@tomponline yes, I'm talking about the possibility to redirect IO on the client side only. The described scenario is the standard shell behaviour when stdout and stderr point to the same file (the terminal), but not when one redirects stderr somewhere else; the expected output on the terminal would be just stdout in such a case. In the lxc's exec scenario I posted, there is no way to do so in the interactive mode, stdout and stderr are combined in a single stream on the server at the time ptys are created unless you enforce non-interactive mode which would create a pipe per the IO stream (and a websocket per stream on the LXD server respectively).

I understand that there are possible workarounds, it was just a bit confusing experience on the client as if I execute the same bash command from a shell, the results will be different, i.e. stderr will go somewhere else but the ptys for stdin and stdout will still be created; what I had in mind is, possibly, having a flag in the `InstanceExecPost` that would allow separating stdout and stderr in the interactive mode. 

This is not something that is blocking for me and if that does not look reasonable now, I'm happy with closing it.
